### PR TITLE
Update README.md to have latest snapshot API changes

### LINKED ISSUE #164: Update csi-driver-host-path README example to use latest images
README needs to be updated to use latest images.  In the current example, snapshot is still at v1alpha1.  Should be updated to use the beta version.

### PR DESCRIPTION
**Pull request by ggriffiths**
Signed-off-by: Grant Griffiths <grant@portworx.com>
**What type of PR is this?**
/kind documentation

**What this PR does / why we need it**:
README had the old snapshotter API as an example

**Which issue(s) this PR fixes**:
Fixes #164

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -183,31 +212,30 @@ An additional way to ensure the driver is working properly is by inspecting the
...
 Status:
   Attached:  true
 Events:      <none>
 ```
 
-
 ## Snapshot support
 
-Since volume snapshot is an alpha feature starting in Kubernetes v1.12, you need to enable feature gate called `VolumeSnapshotDataSource` in the Kubernetes.
+Since volume snapshot is an alpha feature from Kubernetes v1.12 through v1.16, you need to enable feature gate called `VolumeSnapshotDataSource` in Kubernetes for those versions.
```
Comment (xing-yang) [inline README.md]: Change to this:
Volume snapshot was introduced as an alpha feature in Kubernetes v1.12 and it was promoted to beta in Kubernetes v1.16.  Before v1.16, you need to manually enable feature gate called `VolumeSnapshotDataSource` in Kubernetes.

**Comment by xing-yang** [comment]
> * use deploy/latest/deploy-hostpath.sh and add a note which explains which version was used for the output and examples in the rest of the readme.

I agree with this.

**Comment by ggriffiths** [comment]
> The "Deployment" section starts with instructions to run `deploy/kubernetes-1.13/deploy-hostpath.sh`, which isn't in the repo anymore. Please either:
> 
> * replace with the version that you used or
> * use `deploy/latest/deploy-hostpath.sh` and add a note which explains which version was used for the output and examples in the rest of the readme.
> 
> We can't just use "latest" without additional information because if "latest" no longer works as described it would be hard to determine why.

Updated to include this.

File: README.md (hunk 1 above)
Comment (ggriffiths) [reply README.md]: Fixed this.

File: README.md [hunk 2]
```diff
@@ -13,54 +13,70 @@ the cluster as shown below for Kubernetes 1.13. This creates the deployment that
...
 rolebinding.rbac.authorization.k8s.io/csi-provisioner-role-cfg created
-kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-attacher/v1.0.1/deploy/kubernetes/rbac.yaml
+kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-attacher/v2.1.0/deploy/kubernetes/rbac.yaml
 serviceaccount/csi-attacher created
 clusterrole.rbac.authorization.k8s.io/external-attacher-runner created
 clusterrolebinding.rbac.authorization.k8s.io/csi-attacher-role created
 role.rbac.authorization.k8s.io/external-attacher-cfg created
 rolebinding.rbac.authorization.k8s.io/csi-attacher-role-cfg created
-kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v1.0.1/deploy/kubernetes/rbac.yaml
+kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/v2.0.0/deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml
```
Comment (xing-yang) [inline README.md]: We should actually use v2.0.1 because it contains a fix for go module.

File: README.md [hunk 3]
```diff
@@ -13,54 +13,70 @@ the cluster as shown below for Kubernetes 1.13. This creates the deployment that
...
 service/csi-hostpath-provisioner created
 statefulset.apps/csi-hostpath-provisioner created
-   deploy/kubernetes-1.13/hostpath/csi-hostpath-snapshotter.yaml
-        using           image: quay.io/k8scsi/csi-snapshotter:v1.0.1
+   deploy/kubernetes-1.17/hostpath/csi-hostpath-resizer.yaml
+        using           image: quay.io/k8scsi/csi-resizer:v0.4.0
+service/csi-hostpath-resizer created
+statefulset.apps/csi-hostpath-resizer created
+   deploy/kubernetes-1.17/hostpath/csi-hostpath-snapshotter.yaml
+        using           image: quay.io/k8scsi/csi-snapshotter:v2.0.0
```
Comment (xing-yang) [inline README.md]: We should use v2.0.1

**Comment by xing-yang** [comment]
Can you also update the snapshot images to v2.0.1 in the script?

File: README.md (hunk 3 above)
Comment (ggriffiths) [reply README.md]: Fixed - running the same commands from 2.0.1 now to double check the same functioning.

File: README.md (hunk 2 above)
Comment (ggriffiths) [reply README.md]: Fixed this

File: README.md [hunk 4]
```diff
@@ -6,61 +6,94 @@ This repository hosts the CSI Hostpath driver and all of its build and dependent
 - Kubernetes cluster
 - Running version 1.13 or later
 - Access to terminal with `kubectl` installed
+- Snapshot controller and snapshot CRDs applied (see below)
```
Comment (xing-yang) [inline README.md]: This only works for k8s version 1.17 and higher.

@msau42 should we have separate sections for "k8s 1.17 and higher" and "version lower than 1.17" (as we still have 1.16 and 1.15 deployment scripts).

File: README.md [hunk 5]
```diff
@@ -6,61 +6,94 @@ This repository hosts the CSI Hostpath driver and all of its build and dependent
...
+
+## Snapshot controller and CRDs setup
```
Comment (xing-yang) [inline README.md]: This section is only for k8s 1.17 and higher.

File: README.md (hunk 4 above)
Comment (msau42) [reply README.md]: Yes I think that makes sense.

Also, in general, I think this readme is quite long and difficult to find specific topics. Can we potentially break it up a page per example, under a docs/ folder? And make the top-level readme more of TOC?

Comment (ggriffiths) [reply README.md]: Makes sense - I'll update the README to point to point to these two new pages. Will save the old example output for the <1.17 version

Comment (ggriffiths) [reply README.md]: Ok, I've broken up the deployment docs under 
docs/deploy-latest.md
docs/deploy-pre-1.17.md

For the pre-1.17 docs, I copied the old deployment instructions but  reran with deploy/kubernetes-1.16/deploy-hostpath.sh and updated that page.

Comment (ggriffiths) [reply README.md]: oh, I just re-read your comment. I'll break the examples down into separate pages too.

Comment (ggriffiths) [reply README.md]: Ok, broke README.md down to the following files:
```
docs/deploy-latest.md
docs/deploy-pre-1.17.md
docs/example-snasphots.md
docs/example-ephemeral.md
```

File: docs/example-ephemeral.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
+## Inline ephemeral support
+The CSI Hostpath driver (starting with version 1.2.0) now includes support for inline ephemeral volumes. This means that a volume can be specified directly inside a pod spec without the need to use a persistent volume object.
```
Comment (xing-yang) [inline docs/example-ephemeral.md]: Can you add a line here:
Inline ephemeral support was introduced as an alpha feature in Kubernetes 1.15 and was promoted to beta in 1.16.

File: docs/example-snapshots.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+> For Kubernetes < v1.17 
+>
+>  - `$ kubectl apply -f examples/csi-snapshot.yaml`
+> ```
+> $ kubectl apply -f examples/csi-snapshot.yaml
+> volumesnapshot.snapshot.storage.k8s.io/new-snapshot-demo created
+> ```
+>
+> For Kubernetes >= v1.17
```
Comment (xing-yang) [inline docs/example-snapshots.md]: Can you also split example-snapshots.md into two, one for examples < 1.17 and one for 1.17+

File: README.md [hunk 8]
```diff
@@ -6,403 +6,17 @@ This repository hosts the CSI Hostpath driver and all of its build and dependent
...
-      csi:
-        driver: hostpath.csi.k8s.io
-``` 
-
-> See sample YAML file [here](./examples/csi-app-inline.yaml).
-
-Notice the CSI driver is now specified directly in the container spec inside the `volumes:` block.  You can use the [same steps as above][Confirm Hostpath driver works] 
-to verify that the volume has been created and deleted (when the pod is removed).
+Deployment varies depending on the Kubernetes version your cluster is running:
+- [Deployment for Kubernetes 1.17+](docs/deploy-latest.md)
```
[the remark this answers is not in the document]
Comment (ggriffiths) [reply README.md]: Fixed

File: docs/deploy-latest.md [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+For Kubernetes 1.17+, some initial cluster setup is required to install the following:
+- CSI VolumeSnapshot beta CRDs
+- Snapshot Controller
+
+Run the following commands to install these components: 
+```shell
+# Change to the latest supported snapshotter version
+$ SNAPSHOTTER_VERSION=v2.0.1
+
+# Apply snapshotter CRDs
```
[the remark this answers is not in the document]
Comment (ggriffiths) [reply docs/deploy-latest.md]: Fixed

File: docs/deploy-latest.md [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+$ kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/${SNAPSHOTTER_VERSION}/config/crd/snapshot.storage.k8s.io_volumesnapshots.yaml
+
+# Create snapshot controller
+$ kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/${SNAPSHOTTER_VERSION}/deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml
+$ kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/${SNAPSHOTTER_VERSION}/deploy/kubernetes/snapshot-controller/setup-snapshot-controller.yaml
+```
+
+## Deployment
+The easiest way to test the Hostpath driver is to run the `deploy-hostpath.sh` script for the Kubernetes version used by
+the cluster as shown below for Kubernetes 1.13. This creates the deployment that is maintained specifically for that
```
[the remark this answers is not in the document]
Comment (ggriffiths) [reply docs/deploy-latest.md]: Fixed - the output was from 1.17

**Comment by ggriffiths** [comment]
Addressed feedback @xing-yang

File: README.md [hunk 11]
```diff
@@ -6,403 +6,19 @@ This repository hosts the CSI Hostpath driver and all of its build and dependent
...
-      csi:
-        driver: hostpath.csi.k8s.io
-``` 
-
-> See sample YAML file [here](./examples/csi-app-inline.yaml).
-
-Notice the CSI driver is now specified directly in the container spec inside the `volumes:` block.  You can use the [same steps as above][Confirm Hostpath driver works] 
-to verify that the volume has been created and deleted (when the pod is removed).
+Deployment varies depending on the Kubernetes version your cluster is running:
+- [Deployment for Kubernetes 1.17+](docs/deploy-1.17-and-later.md)
```
Comment (xing-yang) [inline README.md]: nit: s/1.17+/1.17 and later

File: README.md [hunk 12]
```diff
@@ -6,403 +6,19 @@ This repository hosts the CSI Hostpath driver and all of its build and dependent
...
+- [Deployment for Kubernetes 1.16 or earlier](docs/deploy-pre-1.17.md)
```
Comment (xing-yang) [inline README.md]: s/1.16 or earlier/1.16 and earlier

File: docs/deploy-1.17-and-later.md [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+
+# Create snapshot controller
+$ kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/${SNAPSHOTTER_VERSION}/deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml
+$ kubectl apply -f https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/${SNAPSHOTTER_VERSION}/deploy/kubernetes/snapshot-controller/setup-snapshot-controller.yaml
+```
+
+## Deployment
+The easiest way to test the Hostpath driver is to run the `deploy-hostpath.sh` script for the Kubernetes version used by
+the cluster as shown below for Kubernetes 1.17. This creates the deployment that is maintained specifically for that
+release of Kubernetes. However, other deployments may also work. For details see the individual READMEs.
```
Comment (xing-yang) [inline docs/deploy-1.17-and-later.md]: This sentence looks odd: "However, other deployments may also work. For details see the individual READMEs."
There is only one README for 1.17 and later, so I don't know what are the other individual READMEs.
Probably better just to drop this sentence.

File: docs/deploy-pre-1.17.md [hunk 14]
```diff
@@ -0,0 +1,210 @@
+## Deployment
+The easiest way to test the Hostpath driver is to run the `deploy-hostpath.sh` script for the Kubernetes version used by
+the cluster as shown below for Kubernetes 1.16. This creates the deployment that is maintained specifically for that
+release of Kubernetes. However, other deployments may also work. For details see the individual READMEs.
```
Comment (xing-yang) [inline docs/deploy-pre-1.17.md]: Same comment here: drop "For details see the individual READMEs." as there is only 1 readme for pre-1.17.

File: docs/deploy-1.17-and-later.md (hunk 13 above)
Comment (ggriffiths) [reply docs/deploy-1.17-and-later.md]: Removed this

File: docs/deploy-pre-1.17.md (hunk 14 above)
Comment (ggriffiths) [reply docs/deploy-pre-1.17.md]: Removed this

File: README.md (hunk 12 above)
Comment (ggriffiths) [reply README.md]: Fixed

File: README.md (hunk 11 above)
Comment (ggriffiths) [reply README.md]: Fixed

File: README.md [hunk 15]
```diff
@@ -6,403 +6,19 @@ This repository hosts the CSI Hostpath driver and all of its build and dependent
 - Kubernetes cluster
 - Running version 1.13 or later
 - Access to terminal with `kubectl` installed
+- For Kubernetes 1.17 or later, the VolumeSnapshot beta CRDs and Snapshot Controller must be installed (see Kubernetes 1.17+ deployment instructions)
```
Comment (msau42) [inline README.md]: clarify this should be installed as part of Kubernetes cluster deployment.

File: docs/deploy-1.17-and-later.md [hunk 16]
```diff
@@ -0,0 +1,248 @@
+## Cluster setup
+For Kubernetes 1.17+, some initial cluster setup is required to install the following:
```
Comment (msau42) [inline docs/deploy-1.17-and-later.md]: Clarify that these steps should only be run if your cluster hasn't already packaged these components in. And maybe add some commands to easily check if these resources already exist.

File: README.md (hunk 15 above)
Comment (ggriffiths) [reply README.md]: Fixed

File: docs/deploy-1.17-and-later.md (hunk 16 above)
Comment (ggriffiths) [reply docs/deploy-1.17-and-later.md]: Updated with some commands to check for the CRDs and Snapshot controller

File: docs/deploy-1.17-and-later.md [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+$ kubectl get volumesnapshotcontents.snapshot.storage.k8s.io
+```
+If any of these commands return the following error message, you must install the corresponding CRD:
+```
+error: the server doesn't have a resource type "volumesnapshotclasses"
+```
+
+Next, check if any pods are running the snapshot-controller image:
+```
+$ kubectl get pods --all-namespaces -o=jsonpath='{range .items[*]}{"\n"}{range .spec.containers[*]}{.image}{", "}{end}{end}' | grep snapshot-controller
```
Comment (msau42) [inline docs/deploy-1.17-and-later.md]: hm this doesn't work on managed distros where you can't see the master pods. Is there another way to check? maybe add a note that if you're running on a managed k8s service, this step may not work, and maybe you'll have to just use the presence of the crds as an indication.

Comment (ggriffiths) [reply docs/deploy-1.17-and-later.md]: Ah yeah, good point. I'm not sure of any other way. @xing-yang any ideas?

I'll add a note to clarify this.

**Comment by msau42** [comment]
/lgtm
/approve
Thanks!

**Pull request merged**